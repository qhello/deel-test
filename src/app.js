import express from "express";
import bodyParser from "body-parser";

import { sequelize } from "./model.js";
import { getProfile } from "./middleware/getProfile.js";
import controllers from "./controllers/index.js";

const app = express();
app.use(bodyParser.json());
app.set("sequelize", sequelize);
app.set("models", sequelize.models);

/**
 * @returns contract by id, only if it belongs to the profile calling
 */
app.get("/contracts/:id", getProfile, controllers.getContractById);

/**
 * @returns contracts that belongs to the profile calling
 */
app.get("/contracts", getProfile, controllers.getContracts);

/**
 * @returns jobs, that are unpaid & belongs to contracts that belongs to the profile calling
 */
app.get("/jobs/unpaid", getProfile, controllers.getJobsUnpaid);

/**
 * @returns 200 OK, or error. Will attempt to pay an unpaid job
 */
app.post("/jobs/:job_id/pay", getProfile, controllers.postJobsPay);

/**
 * @returns 200 OK, or error. Will attempt to deposit specified money into client's balance
 */
app.post(
  "/balances/deposit/:userId",
  getProfile,
  controllers.postBalanceDeposit
);

app.get("/admin/best-profession", controllers.getBestProfession);

app.get("/admin/best-clients", controllers.getBestClients);

export default app;
