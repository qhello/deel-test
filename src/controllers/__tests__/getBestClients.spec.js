import request from "supertest";
import app from "../../app.js";

describe("getBestClients", () => {
  it("should return best clients", async () => {
    const res = await request(app).get("/admin/best-clients").query({
      start: "2020-08-17",
      end: "2020-12-31",
    });

    expect(res.statusCode).toEqual(200);
    expect(res.body).toMatchSnapshot();
    expect(res.body.length).toBe(2);
  });

  it("should return best clients", async () => {
    const res = await request(app).get("/admin/best-clients").query({
      start: "2020-08-10",
      end: "2020-12-31",
      limit: 5,
    });

    expect(res.statusCode).toEqual(200);
    expect(res.body).toMatchSnapshot();
    expect(res.body.length).toBe(4);
  });
});
