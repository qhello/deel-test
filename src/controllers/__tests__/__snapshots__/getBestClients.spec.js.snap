// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBestClients should return best clients 1`] = `
[
  {
    "fullName": "John Snow",
    "id": 3,
    "paid": 200,
  },
  {
    "fullName": "Harry Potter",
    "id": 1,
    "paid": 200,
  },
]
`;

exports[`getBestClients should return best clients 2`] = `
[
  {
    "fullName": "Ash Kethcum",
    "id": 4,
    "paid": 2020,
  },
  {
    "fullName": "Mr Robot",
    "id": 2,
    "paid": 442,
  },
  {
    "fullName": "Harry Potter",
    "id": 1,
    "paid": 442,
  },
  {
    "fullName": "John Snow",
    "id": 3,
    "paid": 200,
  },
]
`;
